### PR TITLE
henry/webrel-1324/setup socket and login

### PR DESCRIPTION
**Pull request by henry-deriv**

### CODE REVIEW

File: src/utils/storage.ts [hunk 1]
```diff
@@ -9,3 +10,50 @@ export const getFromStorage = <T>(key: string): T => {
...
+type TAccountsList = {
+    [k: string]: TAccount &
+        NonNullable<NonNullable<NonNullable<AuthorizeResponse['authorize']>['account_list']>>[number];
+};
+
+/**
+ * Gets the current user `accounts` list from the `localStorage`.
+ */
+export const getAccountsFromLocalStorage = () => {
+    const data = localStorage.getItem('client.accounts');
```
Comment (shayan-deriv) [inline src/utils/storage.ts:21]: we can use `getFromStorage()` which is already in this file.

File: src/utils/storage.ts [hunk 2]
```diff
@@ -9,3 +10,50 @@ export const getFromStorage = <T>(key: string): T => {
...
+    if (!data) return;
+    // Cast parsed JSON data to infer return type
+    return JSON.parse(data) as TAccountsList;
+};
+
+/**
+ * Gets the active `loginid` for the current user from the `localStorage`.
+ */
+export const getActiveLoginIdFromLocalStorage = () => {
+    const active_loginid = localStorage.getItem('active_loginId');
```
Comment (shayan-deriv) [inline src/utils/storage.ts:33]: same here ( `getFromStorage()` )

File: src/utils/storage.ts [hunk 3]
```diff
@@ -9,3 +10,50 @@ export const getFromStorage = <T>(key: string): T => {
...
+
+    return active_loginid;
+};
+
+export const getActiveAuthToken = () => {
+    const accounts = getAccountsFromLocalStorage();
+    const active_loginid = getActiveLoginIdFromLocalStorage();
+    const token_from_query_params = getActiveAuthTokenFromQueryParameters();
+    // If there is no active loginid or no accounts list, return undefined.
+    // if (!active_loginid || !accounts) return;
```
Comment (shayan-deriv) [inline src/utils/storage.ts]: can we remove these if we don't need them?

File: src/api/hooks/useAuthorize.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        payload: { authorize: current_token || '' },
+        options: { enabled: Boolean(current_token) },
+    });
+
+    // Add additional information to the authorize response.
+    const modified_authorize = useMemo(() => ({ ...data?.authorize }), [data?.authorize]);
+
+    const switchAccount = useCallback(
+        (loginid: string) => {
+            const active_loginid = localStorage.getItem('active_loginId');
```
Comment (shayan-deriv) [inline src/api/hooks/useAuthorize.ts:23]: same here ( `getFromStorage()` ). if you think we don't need `getFromStorage()` maybe we can remove it.

File: src/api/hooks/useAuthorize.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            if (active_loginid !== loginid) {
+                localStorage.setItem('active_loginid', loginid);
+                invalidate('authorize');
```
Comment (shayan-deriv) [inline src/api/hooks/useAuthorize.ts:26]: I have a question since I haven't work much with react-query. what will happen exactly when we call invlidate ?

Comment (shayan-deriv) [reply src/api/hooks/useAuthorize.ts:26]: it will refetch data for the new `loginid`?

File: src/utils/storage.ts (hunk 2 above)
Comment (henry-deriv) [reply src/utils/storage.ts:33]: We should just remove getFromStorage lol. It's abstracting a simple line of code for no reason.

Comment (shayan-deriv) [reply src/utils/storage.ts:33]: I agree. please remove

File: tsconfig.json [hunk 6]
```diff
@@ -33,5 +33,6 @@
     }
```
Comment (niloofar-deriv) [inline tsconfig.json:35]:
```suggestion
    "Types/*": ["types/*"]
    }
```

You need to add this path

File: src/App.tsx [hunk 7]
```diff
@@ -1,13 +1,16 @@
-import ErrorBoundary from 'Components/common/error-boundary';
-import Layout from 'Components/layout';
-import AuthProvider from 'Contexts/authProvider';
+import ErrorBoundary from './components/common/error-boundary';
+import Layout from './components/layout';
+import useLogin from './api/hooks/useLogin';
+import React from 'react';
```
Comment (niloofar-deriv) [inline src/App.tsx]: Do we need importing React?

File: src/App.tsx [hunk 8]
```diff
@@ -1,13 +1,16 @@
-import ErrorBoundary from 'Components/common/error-boundary';
-import Layout from 'Components/layout';
-import AuthProvider from 'Contexts/authProvider';
+import ErrorBoundary from './components/common/error-boundary';
```
Comment (niloofar-deriv) [inline src/App.tsx]:
```suggestion
import ErrorBoundary from 'Components/common/error-boundary';
```

File: src/App.tsx [hunk 9]
```diff
@@ -1,13 +1,16 @@
...
+import Layout from './components/layout';
```
Comment (niloofar-deriv) [inline src/App.tsx]:
```suggestion
import Layout from 'Components/layout';
```

File: src/App.tsx [hunk 10]
```diff
@@ -1,13 +1,16 @@
...
+import useLogin from './api/hooks/useLogin';
```
Comment (niloofar-deriv) [inline src/App.tsx]:
```suggestion
import useLogin from 'Api/hooks/useLogin';
```

Can we add this to our vite and ts config, so we can import it like this

File: src/App.tsx [hunk 11]
```diff
@@ -1,13 +1,16 @@
...
 
-const App = () => (
-    <AuthProvider>
+const App = () => {
+    const { is_logged_in } = useLogin();
+
+    return (
         <ErrorBoundary>
-            <Layout>App</Layout>
+            <Layout is_logged_in={is_logged_in}>App</Layout>
```
Comment (niloofar-deriv) [inline src/App.tsx]: Why we are passing is_logged_in as a props? can't we just use this custom hook inside the layout?

Comment (niloofar-deriv) [reply src/App.tsx]: IMO we don't need to pass it as a props.

Comment (niloofar-deriv) [reply src/App.tsx]: Also it will increase the code complexity

File: package.json [hunk 12]
```diff
@@ -16,16 +16,21 @@
...
+    "@deriv/api-types": "^1.0.126",
+    "@deriv/deriv-api": "^1.0.13",
     "@radix-ui/react-dialog": "^1.0.5",
     "@radix-ui/react-icons": "^1.3.0",
-    "@radix-ui/react-slot": "^1.0.2",
+    "@radix-ui/react-slot": "^1.0.2",  
     "class-variance-authority": "^0.7.0",
     "clsx": "^2.0.0",
-    "lint-staged": "^15.0.1",
+    "lint-staged": "^15.0.1",
```
Comment (hasan-deriv) [inline package.json]: Shouldn't it be dev dependency?

**Review by hasan-deriv — APPROVED** [review]

File: src/App.tsx (hunk 7 above)
Comment (henry-deriv) [reply src/App.tsx]: Either linter or pre commit checks were complaining i believe.

File: src/App.tsx (hunk 11 above)
Comment (henry-deriv) [reply src/App.tsx]: yup can remove

File: package.json (hunk 12 above)
Comment (henry-deriv) [reply package.json]: I pulled master and it came into my PR. I dont remember who added it. @shayan-deriv @niloofar-deriv  ?

Comment (shayan-deriv) [reply package.json]: yes should be devDep. 
cc: @niloofar-deriv

Comment (henry-deriv) [reply package.json]: ok i will resolve it here

File: src/utils/websocket/config.ts [hunk 13]
```diff
@@ -1,3 +1,67 @@
-const config = () => {};
+export const domain_app_ids = {
```
Comment (matin-deriv) [inline src/utils/websocket/config.ts]: Can we move this to a constant file!

File: src/utils/websocket/config.ts [hunk 14]
```diff
@@ -1,3 +1,67 @@
...
+        // TODO: [app-link-refactor] - Remove backwards compatibility for `deriv.app`
+        window.localStorage.removeItem('config.default_app_id');
+        app_id = 22169;
+    } else if (/staging-smarttrader\.deriv\.com/i.test(window.location.hostname)) {
+        window.localStorage.removeItem('config.default_app_id');
+        app_id = 22169;
+    } else if (/staging-smarttrader\.deriv\.app/i.test(window.location.hostname)) {
+        // TODO: [app-link-refactor] - Remove backwards compatibility for `deriv.app`
+        window.localStorage.removeItem('config.default_app_id');
+        app_id = 22169;
```
Comment (matin-deriv) [inline src/utils/websocket/config.ts]: What do you think about accessing the app_ids from a constant file which is a single source of truth?

File: src/utils/utility.ts [hunk 15]
```diff
@@ -0,0 +1,83 @@
+import useAuthorize from '../api/hooks/useAuthorize';
```
Comment (matin-deriv) [inline src/utils/utility.ts]:
```suggestion
import useAuthorize from 'Api/hooks/useAuthorize';
```

File: tsconfig.json [hunk 16]
```diff
@@ -21,18 +21,20 @@
...
-    },
-    "include": ["src", "stories/*.stories.tsx", "stories"],
-    "references": [{ "path": "./tsconfig.node.json" }]
+    /* Absoloute path */
+    "baseUrl": "./src/",
+    "paths": {
+      "Components/*": ["components/*"],
+      "Constants/*": ["constants/*"],
+      "Contexts/*": ["contexts/*"],
+      "Hooks/*": ["hooks/*"],
```
Comment (matin-deriv) [inline tsconfig.json]: Do you need to replace this with `Api`?

File: tsconfig.json [hunk 17]
```diff
@@ -21,18 +21,20 @@
...
-            "Contexts/*": ["contexts/*"],
-            "Hooks/*": ["hooks/*"],
-            "Utils/*": ["utils/*"],
-            "Styles/*": ["styles/*"],
-            "Translations/*": ["translations/*"]
-        }
-    },
-    "include": ["src", "stories/*.stories.tsx", "stories"],
-    "references": [{ "path": "./tsconfig.node.json" }]
+    /* Absoloute path */
```
Comment (matin-deriv) [inline tsconfig.json]:
```suggestion
    /* Absolute path */
```

File: src/utils/websocket/config.ts [hunk 18]
```diff
@@ -1,3 +1,67 @@
...
+    const search = window.location.search;
+    if (search) {
+        const params = new URLSearchParams(document.location.search.substring(1));
+        active_loginid_from_url = params.get('acct1');
+    }
+
+    const loginid = window.localStorage.getItem('active_loginid') || active_loginid_from_url;
+    const is_real = loginid && !/^(VRT|VRW|VRTC)/.test(loginid);
+
+    const server = is_real ? 'green' : 'blue';
```
Comment (matin-deriv) [inline src/utils/websocket/config.ts]: Would be nice to move the servers to a constant file as well 🙏🏼

File: src/utils/websocket/APIProvider.tsx [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    return window.ReactQueryClient;
+};
+
+// This is a temporary workaround to share a single `DerivAPIBasic` instance for every unique URL.
+// Later once we have each package separated we won't need this anymore and can remove this.
+const getDerivAPIInstance = (): DerivAPIBasic => {
+    const endpoint = getSocketURL();
+    const language = 'EN'; // Need to use the language from the app context.
```
Comment (matin-deriv) [inline src/utils/websocket/APIProvider.tsx]: You can import `DEFAULT_LANGUAGE` from the constant file `translations.ts`

**Review by matin-deriv — CHANGES_REQUESTED** [review]

File: tsconfig.json (hunk 16 above)
Comment (henry-deriv) [reply tsconfig.json]: ah i thought i did

File: src/utils/websocket/config.ts (hunk 14 above)
Comment (henry-deriv) [reply src/utils/websocket/config.ts]: 👍

File: src/utils/login.ts [hunk 20]
```diff
@@ -0,0 +1,16 @@
+import { getTopLevelDomain } from './utility';
+import { getAppId } from './websocket/config';
+
+export const loginUrl = () => {
+    const server_url = localStorage.getItem('config.server_url');
+    const language = 'en';
```
Comment (matin-deriv) [inline src/utils/login.ts]: Is this always `en`? should we get this from the localstorage or use the default language?

File: src/components/layout/header/login-logout/LoginLogoutButton.tsx [hunk 21]
```diff
@@ -0,0 +1,7 @@
+import { loginUrl } from 'Utils/login';
+
+const LoginLogoutButton = () => {
+    return <a href={loginUrl()}>login</a>;
```
Comment (maryia-deriv) [inline src/components/layout/header/login-logout/LoginLogoutButton.tsx]: @henry-deriv  shouldn't we use the Button component from components/ui here? :)

**Review by maryia-deriv — CHANGES_REQUESTED** [review]
just one question :)

File: src/components/layout/header/login-logout/LoginLogoutButton.tsx (hunk 21 above)
Comment (matin-deriv) [reply src/components/layout/header/login-logout/LoginLogoutButton.tsx]: @maryia-deriv I believe this is just a temp button to test the login and logout, we will implement the header later on! still a WIP 😄

File: src/utils/login.ts (hunk 20 above)
Comment (henry-deriv) [reply src/utils/login.ts]: Actually language change is not handled yet. Will do when we have language switcher. Will pull from local storage but if local storage value is empty we'll default to DEFAULT_LANGUAGE

**Review by maryia-deriv — APPROVED** [review]

**Review by matin-deriv — DISMISSED** [review]

File: src/api/hooks/utils.ts [hunk 22]
```diff
@@ -0,0 +1,23 @@
+export const getQueryKeys = (name: string, props?: Record<string, unknown>) => {
```
Comment (niloofar-deriv) [inline src/api/hooks/utils.ts:1]:
```suggestion
export const getQueryKeys = <T extends object>(name: string, props?: Record<string, T>) => {
```

🤔

File: src/api/hooks/useResponsive.ts
Comment (niloofar-deriv) [inline src/api/hooks/useResponsive.ts:1]: This is not related to the apis hook, I guess it's better to remove the file, later if we need that we can recreate it under `src/hooks` folder!

File: src/api/hooks/utils.ts (hunk 22 above)
Comment (henry-deriv) [reply src/api/hooks/utils.ts:1]: doesnt work.

**Review by niloofar-deriv — APPROVED** [review]

**Review by matin-deriv — DISMISSED** [review]

**Review by ako-deriv — CHANGES_REQUESTED** [review]
The CI is failing due to wrong type definition/usage.
please check it.

File: src/api/hooks/useLogin.ts [hunk 23]
```diff
@@ -0,0 +1,31 @@
+import { deleteQueryParams, readLoginQueryParams } from 'Utils/url';
+import useAuthorize from './useAuthorize';
+import { storeClientAccounts } from 'Utils/utility';
+import { useEffect, useMemo } from 'react';
+
+const useLogin = () => {
+    const search = window.location.search;
+    const client_account = readLoginQueryParams();
```
Comment (shayan-deriv) [inline src/api/hooks/useLogin.ts]:
```suggestion
    const client_account_params = readLoginQueryParams();
```
just for being more clear

File: src/api/hooks/useLogin.ts [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+    const { data: account_list } = useAuthorize();
```
Comment (shayan-deriv) [inline src/api/hooks/useLogin.ts]:
```suggestion
    const { data: client_info } = useAuthorize();
```
same here

File: src/utils/utility.ts [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+    account_list?.forEach(account => {
+        Object.keys(account).forEach(param => {
+            if (param === 'loginid') {
+                if (!active_loginid && !account.is_disabled) {
+                    if (!account.is_virtual) {
+                        active_loginid = account[param] ?? '';
+                    } else if (account.is_virtual) {
+                        // TODO: [only_virtual] remove this to stop logging non-SVG clients into virtual
+                        active_loginid = account[param] ?? '';
+                    }
```
Comment (shayan-deriv) [inline src/utils/utility.ts]:
```suggestion
                   active_loginid = account[param] ?? '';
```
can we change it for now? 
I see that there is a TODO for the `else if` block so we can add more logics but for now I think we can remove it. what do u think @henry-deriv ?

Comment (henry-deriv) [reply src/utils/utility.ts]: @shayan-deriv i changed it. Take a look!

File: src/api/hooks/useMutation.ts
Comment (shayan-deriv) [inline src/api/hooks/useMutation.ts:1]: what are we using this for ? 🤔  @henry-deriv

File: src/utils/utility.ts (hunk 25 above)
Comment (shayan-deriv) [reply src/utils/utility.ts]: thanks man. LGTM

File: src/api/hooks/useMutation.ts
Comment (shayan-deriv) [reply src/api/hooks/useMutation.ts:1]: ignore my question 😄 figured it out

**Review by shayan-deriv — APPROVED** [review]

**Review by matin-deriv — APPROVED** [review]

**Review by niloofar-deriv — APPROVED** [review]

**Pull request merged**